PCSC-lite bindings

TheBestTvarynka opened the pull request:
Hi,
I've implemented Rust bindings to the [`pcsc-lite`](https://pcsclite.apdu.fr/) library in this pull request.

This pull request is the first in a series of pull requests for system smart card support. To support system-provided smart cards, we need to somehow access them from `sspi-rs`. On Windows, we will use native WinSCard API. We plan to use the `pcsc-lite` library on Linux and macOS; to do so, we need these bindings.

_**Why can't we just use the [`pcsc-sys`](https://crates.io/crates/pcsc-sys) crate?**_

The main reason is that it links with the PCSC framework on macOS which was deprecated for years. Now Apple encourages developers to use their _CryptoTokenKit_.
I observed the `pcsc-lite` API and decided that it's easier to write our bindings than fork, patch, and maintain the `pcsc-sys` crate. I've reused some types from our `ffi-types` crate. It made my work easier.

pacmancoder commented on crates/pcsc-lite-rs/build.rs:
```diff
@@ -0,0 +1,18 @@
+use pkg_config::Config;
+
+fn main() {
+    if let Ok(lib_dir) = env::var("PCSC_LIB_DIR") {
+        println!("cargo:rustc-link-search=native={}", lib_dir);
+        println!(
+            "cargo:rustc-link-lib={}",
+            env::var("PCSC_LIB_NAME").unwrap_or_else(|_| "pcsclite".to_string())
+        );
+    } else {
+        Config::new()
+            .atleast_version("1")
+            .probe("libpcsclite")
+            .expect("Could not find a PCSC library");
+    }
```
To clarify -- will this library be statically linked to the resulting executable?

TheBestTvarynka replied:
Nope, it'll be dynamically linked.
The `pcsc-lite` API is stable and likely will not be changed. So, I decided to link it dynamically

pacmancoder replied:
Hm... I have a concern that when we start to use this in IronRDP or other products, 90% of the time this library will be absent on the system, and the application will simply fail to start because of that, even if the user does not need smart card auth support at all.

Maybe it is better to use `dlopen`/`dlsym` machinery for smart card auth to load `pcsc-lite` on runtime? Not only this will make `pcsc-lite` presence on the machine optional, but this will also simplify the compilation process of sspi-rs (no need for installing/compiling `pcsc-lite` on dev machine)

Alternatively, we could just statically link the library.

cc @CBenoit @awakecoding

TheBestTvarynka replied:
> I wonder If we actually had any problems I described above 🤔

we used the `pcsc` crate only on Windows. So, the WinSCard API was always with us on Windows 😅

awakecoding replied:
> Maybe it is better to use dlopen/dlsym machinery for smart card auth to load pcsc-lite on runtime? Not only this will make pcsc-lite presence on the machine optional, but this will also simplify the compilation process of sspi-rs (no need for installing/compiling pcsc-lite on dev machine)

static linking is not really an option for pcsc-lite, and yes, adding a runtime dependency that is likely to be missing is a concern, especially on Linux. macOS should not be much of an issue issue though, even if I'd prefer having runtime loading of the DLL. Maybe we can begin by enabling the feature at build time only for macOS, and then rework this to find and load the pcsc-lite library at runtime in a second time before trying to use it on Linux?

TheBestTvarynka replied:
> even if I'd prefer having runtime loading of the DLL

got it. I'll update the PR tomorrow.
I already implemented a similar behavior during the dev testing. Thus it's not a big problem to add it now

TheBestTvarynka replied:
> I'll update the PR tomorrow

done

pacmancoder commented on crates/pcsc-lite-rs/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "pcsc-lite-rs"
```
Currently `pcsc-lite-rs` is only providing low-level bindings to the native `pcsc-lite` library, and therefore it feels right to use `*-sys` suffix for such crates, even if a high-level counterpart is not planned.

Maybe @CBenoit or @awakecoding will have some thoughts about that?

TheBestTvarynka replied:
:thinking: 
I think we can name it `pcsc-lite-sys`. What do you think?

pacmancoder replied:
Sounds good 👍

TheBestTvarynka commented:
@pacmancoder @awakecoding 

I rewrote the bindings. Now it loads the library in the runtime. I decided to move the code into the `ffi/src/winscard/pcsc-lite`. I thing we don't need a separate crate for just type definitions and one load function.

Also, I refactored the `sspi` and now it can be compiled on macOS/Linux with the `scard` feature turned on.

cc @CBenoit

pacmancoder reviewed (approved):
Looks good to me now! 🥇

TheBestTvarynka commented:
@pacmancoder I think it's better to _squash and merge_ instead of _merge_ for such intermediate PRs (even if we merge into the dev branch). It'll allow us to have one good-looking final PR which consists of all these intermediate PRs (for example, #210 )

pacmancoder commented:
@TheBestTvarynka I always squash & merge when merging to master, I made a simple merge because I thought it would be simpler for you to manage dev branch bisect changes, etc., you could squash them by yourself in the dev branch if this were an intention. I'll squash & merge to Apriorit dev branches from now on
